pkg/common/cns-lib/volume/util.go

```diff
@@ -161,3 +166,154 @@ func updateQueryResult(ctx context.Context, m *defaultManager, res *cnstypes.Cns
 	}
 	return res
 }
+
+// setupConnection connects to CNS and updates the VSphereUser to the session's user.
+func setupConnection(ctx context.Context, virtualCenter *cnsvsphere.VirtualCenter,
+	spec *cnstypes.CnsVolumeCreateSpec) error {
+	log := logger.GetLogger(ctx)
+	// Set up the VC connection.
+	err := virtualCenter.ConnectCns(ctx)
+	if err != nil {
+		log.Errorf("ConnectCns failed with err: %+v", err)
+		return err
+	}
+	// If the VSphereUser in the CreateSpec is different from session user,
+	// update the CreateSpec.
+	s, err := virtualCenter.Client.SessionManager.UserSession(ctx)
+	if err != nil {
+		log.Errorf("failed to get usersession with err: %v", err)
+		return err
+	}
+	if s.UserName != spec.Metadata.ContainerCluster.VSphereUser {
+		log.Debugf("Update VSphereUser from %s to %s", spec.Metadata.ContainerCluster.VSphereUser, s.UserName)
+		spec.Metadata.ContainerCluster.VSphereUser = s.UserName
+	}
+	return nil
+}
+
+// getPendingCreateVolumeTaskFromMap returns the CreateVolume task for a volume stored in the volumeTaskMap.
+func getPendingCreateVolumeTaskFromMap(ctx context.Context, volNameFromInputSpec string) *object.Task {
+	var task *object.Task
+	log := logger.GetLogger(ctx)
+	taskDetailsInMap, ok := volumeTaskMap[volNameFromInputSpec]
+	if ok {
+		task = taskDetailsInMap.task
+		log.Infof("CreateVolume task still pending for Volume: %q, with taskInfo: %+v",
+			volNameFromInputSpec, task)
+	}
+	return task
+}
+
+// invokeCNSCreateVolume truncates the input volume name and invokes a CreateVolume operation for that volume on CNS.
+func invokeCNSCreateVolume(ctx context.Context, virtualCenter *cnsvsphere.VirtualCenter,
+	spec *cnstypes.CnsVolumeCreateSpec) (*object.Task, error) {
+	var cnsCreateSpecList []cnstypes.CnsVolumeCreateSpec
+	log := logger.GetLogger(ctx)
+	// Truncate the volume name to make sure the name is within 80 characters before calling CNS.
+	if len(spec.Name) > maxLengthOfVolumeNameInCNS {
+		volNameAfterTruncate := spec.Name[0 : maxLengthOfVolumeNameInCNS-1]
+		log.Infof("Create Volume with name %s is too long, truncate it to %s", spec.Name, volNameAfterTruncate)
```

Comment:
Can you paste the logs of a create volume workflow when run in PRODUCTION log level? I'm wondering if we have too much logging.

Reply:
For a simple create without any errors/retries:
```
# kubectl create -f pvc.yaml 
persistentvolumeclaim/example-vanilla-block-pvc created
# kubectl get pvc
NAME                        STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS               AGE
example-vanilla-block-pvc   Bound    pvc-4d4310fd-a7f8-4807-89e6-abc2bb7ea875   1Mi        RWO            example-vanilla-block-sc   18s
# kubectl get cnsvolumeoperationrequest -A
NAMESPACE           NAME                                       AGE
vmware-system-csi   pvc-4d4310fd-a7f8-4807-89e6-abc2bb7ea875   54s
```

These are the logs:

```
{"log":"{\"level\":\"info\",\"time\":\"2021-06-02T19:07:24.715884403Z\",\"caller\":\"vanilla/controller.go:659\",\"msg\":\"CreateVolume: called with args {Name:pvc-4d4310fd-a7f8-4807-89e6-abc2bb7ea875 CapacityRange:required_bytes:1048576  VolumeCapabilities:[mount:\u003cfs_type:\\\"ext4\\\" \u003e access_mode:\u003cmode:SINGLE_NODE_WRITER \u003e ] Parameters:map[] Secrets:map[] VolumeContentSource:\u003cnil\u003e AccessibilityRequirements:\u003cnil\u003e XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}\",\"TraceId\":\"541e8403-9c9e-40d8-a716-9b8662d4f16b\"}\n","stream":"stderr","time":"2021-06-02T19:07:24.718439537Z"}
{"log":"{\"level\":\"info\",\"time\":\"2021-06-02T19:07:37.714439211Z\",\"caller\":\"volume/manager.go:298\",\"msg\":\"CreateVolume: VolumeName: \\\"pvc-4d4310fd-a7f8-4807-89e6-abc2bb7ea875\\\", opId: \\\"61ab2196\\\"\",\"TraceId\":\"541e8403-9c9e-40d8-a716-9b8662d4f16b\"}\n","stream":"stderr","time":"2021-06-02T19:07:37.717735093Z"}
{"log":"{\"level\":\"info\",\"time\":\"2021-06-02T19:07:37.726958117Z\",\"caller\":\"volume/util.go:311\",\"msg\":\"Volume created successfully. VolumeName: \\\"pvc-4d4310fd-a7f8-4807-89e6-abc2bb7ea875\\\", volumeID: {{} \\\"7603df21-b8f9-404f-8a34-ef67d33b6b87\\\"}\",\"TraceId\":\"541e8403-9c9e-40d8-a716-9b8662d4f16b\"}\n","stream":"stderr","time":"2021-06-02T19:07:37.727383122Z"}
```